sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/tuplesketchesAggregates.scala

```diff
@@ -0,0 +1,1895 @@
+}
+case class IntersectionTupleAggregationBuffer[S <: Summary](intersection: Intersection[S])
+    extends TupleSketchState[S] {
+  override def serialize(): Array[Byte] = intersection.getResult.toByteArray
+  override def eval(): Array[Byte] = intersection.getResult.toByteArray
+}
+case class FinalizedTupleSketch[S <: Summary](sketch: Sketch[S]) extends TupleSketchState[S] {
+  override def serialize(): Array[Byte] = sketch.toByteArray
+  override def eval(): Array[Byte] = sketch.toByteArray
+}
+
+/**
+ * The TupleSketchAggDouble function utilizes a Datasketches TupleSketch instance to count a
+ * probabilistic approximation of the number of unique values in a given column with associated
+ * double type summary values, and outputs the binary representation of the TupleSketch.
+ *
+ * See [[https://datasketches.apache.org/docs/Tuple/TupleSketches.html]] for more information.
+ *
+ * @param key
+ *   key expression against which unique counting will occur
+ * @param summary
+ *   summary expression (double type) against which different mode aggregations will occur
+ * @param lgNomEntries
+ *   the log-base-2 of nomEntries decides the number of buckets for the sketch
+ * @param mode
+ *   the aggregation mode for numeric summaries (sum, min, max, alwaysone)
+ * @param mutableAggBufferOffset
+ *   offset for mutable aggregation buffer
+ * @param inputAggBufferOffset
+ *   offset for input aggregation buffer
+ */
+case class TupleSketchAggDouble(
+    key: Expression,
+    summary: Expression,
+    lgNomEntries: Expression,
+    mode: Expression,
+    override val mutableAggBufferOffset: Int,
+    override val inputAggBufferOffset: Int)
+    extends TupleSketchAggBase[java.lang.Double, DoubleSummary]
+    with QuaternaryLike[Expression]
+    with SummaryAggregateMode {
+
+  // Constructors
+  def this(key: Expression, summary: Expression) = {
+    this(
+      key,
+      summary,
+      Literal(ThetaSketchUtils.DEFAULT_LG_NOM_LONGS),
+      Literal(ThetaSketchUtils.MODE_SUM),
+      0,
+      0)
+  }
+
+  def this(key: Expression, summary: Expression, lgNomEntries: Expression) = {
+    this(key, summary, lgNomEntries, Literal(ThetaSketchUtils.MODE_SUM), 0, 0)
+  }
+
+  def this(key: Expression, summary: Expression, lgNomEntries: Expression, mode: Expression) = {
+    this(key, summary, lgNomEntries, mode, 0, 0)
+  }
+
+  /**
+   * Override inputTypes to specify key, summary (double/float), lgNomEntries (int), and mode
+   * (string) parameters.
+   */
+  override def inputTypes: Seq[AbstractDataType] =
+    Seq(
+      keyInputTypes,
+      summaryInputType,
+      IntegerType,
+      StringTypeWithCollation(supportsTrimCollation = true))
+
+  /**
+   * Override checkInputDataTypes to validate base inputs (key, summary, lgNomEntries) and mode
+   * parameter.
+   */
+  override def checkInputDataTypes(): TypeCheckResult = {
+    val defaultCheck = checkBaseInputDataTypes()
+    if (defaultCheck.isFailure) return defaultCheck
```

Comment:
Let's not use `return` in any of this Scala because its behavior is somewhat unpredictable in the context of lambdas, e.g. Catalyst rules. Even if the code usage isn't in a Catalyst rule, it's best to be consistent and avoid this, it has lead to difficult surprises in the past when code is added or changed. Here and elsewhere in the PR.